### ISSUE TITLE
How to contribute from your Fork for Labs

### ISSUE THREAD

**Issue by soda4fries** [soda4fries/WIA1002_S2_24_occ9#17]
create a tutorial for the forking and sumitting lab works

**Comment by Nas-Mohd** [comment]
## Forking a Repository
Before we can make changes, or rather propose changes to the repository, we must first make a fork of the repository to serve as our copy. We will be making our changes to the fork then making pull requests to try and update the repository based on our changes.

> A fork is a new repository that shares code and visibility settings with the original “upstream” repository.
1.  Go to the repository page and click the fork button

![image_2024-04-17_231658681](https://github.com/soda4fries/WIA1002_S2_24_occ9/assets/142051728/e085f30c-be60-4f2c-bc70-b4b6a50ef50c)

2. You will be directed to a page to confirm and edit the fork of the repository. Just hit the "Create Fork" button at the bottom of the page.
![image_2024-04-17_232326443](https://github.com/soda4fries/WIA1002_S2_24_occ9/assets/142051728/43834471-6e0d-4c91-8270-72eaf433869c)


3. Wait for a bit then you will now have your own fork of the repository.

## Making Changes and Pull Requests

With the fork that we now have, we can make changes to the fork and try to push them into the original repository by way of making pull requests.

> A pull request is a proposal to merge a set of changes from one branch into another. In a pull request, collaborators can review and discuss the proposed set of changes before they integrate the changes into the main codebase.

### Make sure your computer has [Git](https://git-scm.com/downloads) installed.

1.  There are two ways to go about editing and making changes to the fork.
 - Through GitHub. GitHub allows you to upload files and create new files directly.
![image_2024-04-17_234445198](https://github.com/soda4fries/WIA1002_S2_24_occ9/assets/142051728/523dba88-a2a7-4556-a355-9c9b177f6c57)

 - Through your own terminal or IDE. VSCode, IntelliJ, NetBeans and all the other common IDEs have Git and version control functionalities integrated that you can use to clone your fork to make changes. 

2. For your initial pull request, head to the "Labs" directory and create a new folder named after your matric number. In GitHub, you can do the following:
- Click 'Add File' then 'Create New File'
![image](https://github.com/soda4fries/WIA1002_S2_24_occ9/assets/142051728/f92f08d1-d2c2-4322-a9a0-aac5e5ae9ea3)
-  Name the file "(Matric Number)/hi.txt". It will look something like this:
![image](https://github.com/soda4fries/WIA1002_S2_24_occ9/assets/142051728/fb61ae5f-aae0-4624-907a-a85927dfe177)
- Type 'hi' or anything you want in the file.
![image](https://github.com/soda4fries/WIA1002_S2_24_occ9/assets/142051728/16f76c51-eb1e-47bd-8cdf-5dd341a21770)

- Hit 'Commit changes'
![image](https://github.com/soda4fries/WIA1002_S2_24_occ9/assets/142051728/4b99b517-8427-495d-adb8-cef0034e5993)
- A window will appear. You can edit the commit message (Protip: make it meaningful :>), then you can hit 'Commit changes'.
![image](https://github.com/soda4fries/WIA1002_S2_24_occ9/assets/142051728/869b9dc4-bd97-4d02-9b07-b64823e7552e)

4. After making changes and committing, you can now open a pull request. Go to your fork repository page. If you notice, there is a text now saying that "This branch is X commits ahead of (upstream repository). You can either click the "X commits ahead of" hypertext or the 'Contribute' button and hit 'Open pull request'
![image](https://github.com/soda4fries/WIA1002_S2_24_occ9/assets/142051728/07b2f7dc-71fe-4ea9-bb35-fae5cf41a34d)

5. You will be directed to a page that shows which branch will be requested to be pulled to what other branch. The differences/changes will also be shown down below, but honestly just hit 'Create pull request' (cause i'm also still learning this stuff 💀 )
![image](https://github.com/soda4fries/WIA1002_S2_24_occ9/assets/142051728/0b6da732-0a7d-4cb4-8592-0bad13a5875a)

6. Lastly add a title for the pull request and hit 'Create pull request'
![image](https://github.com/soda4fries/WIA1002_S2_24_occ9/assets/142051728/61185a54-5edb-4b96-a22e-94eea76cf62f)

Congrats you have now made a pull request and will be added to be a collaborator soon ^^

## Fetching and Syncing Your Fork

Now that you are able to make changes to the repository by updating your fork and making pull requests, there is another thing to keep track of. The original repository can and will be updated and these changes will not be automatically added to your fork. So you will need to 'fetch' these new changes into your own fork or copy of the repo.

> Fetching is used to retrieve new work done by other people.

### Using GitHub

- Honestly just read [this](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork)
tl:dr just hit 'Sync Fork' and 'Update branch' then pray that there aren't any conflicts :p

### Using a terminal

1. Navigate first to your repository in the terminal if you aren't in it. Then run the following codes:
` git remote add upstream https://github.com/soda4fries/WIA1002_S2_24_occ9`
`git fetch upstream`
`git checkout main`
`git merge upstream/main`

> Tip: Syncing your fork only updates your local copy of the repository. To update your fork on GitHub.com, you must [push your changes](https://docs.github.com/en/get-started/using-git/pushing-commits-to-a-remote-repository).
